kernel/src/listed_log_files.rs

```diff
@@ -41,31 +41,84 @@ pub(crate) struct ListedLogFiles {
-/// `end_version` is not specified, files up to the most recent version will be included.
+/// (inclusive) and `end_version` (inclusive) taking into account the `log_tail` which was
+/// (ostentibly) returned from the catalog. Note that the `log_tail` must strictly adhere to being
+/// a 'tail' - that is, it goes from X..Y (inclusive) and Y must be the latest version of the
+/// table. Note that if it overlaps with commits listed from the filesystem, the `log_tail` will
+/// take precedence.
 ///
-/// Note: this calls [`StorageHandler::list_from`] to get the list of log files.
+/// [`ParsedLogPath`] may be a commit or a checkpoint. If `start_version` is not specified, the
+/// files will begin from version number 0. If `end_version` is not specified, files up to the most
+/// recent version will be included.
+///
+/// Note: this may call [`StorageHandler::list_from`] to get the list of log files unless the
+/// provided log_tail covers the entire requested range.
+///
+/// TODO: consider renaming? `maybe_list_log_files`? `list_log_files_with_tail`?
 fn list_log_files(
     storage: &dyn StorageHandler,
     log_root: &Url,
+    log_tail: Vec<ParsedLogPath>,
     start_version: impl Into<Option<Version>>,
     end_version: impl Into<Option<Version>>,
 ) -> DeltaResult<impl Iterator<Item = DeltaResult<ParsedLogPath>>> {
+    #[cfg(debug_assertions)]
+    {
+        // log_tail is only commits
+        for entry in &log_tail {
+            debug_assert!(
+                matches!(entry.file_type, LogPathFileType::Commit),
+                "log_tail should only contain commits, found {:?}",
+                entry.file_type
+            );
+        }
+        // log_tail has no gaps nor duplicates
+        if !log_tail.is_empty() {
+            for i in 1..log_tail.len() {
+                debug_assert_eq!(
+                    log_tail[i].version,
+                    log_tail[i - 1].version + 1,
+                    "log_tail should be contiguous without gaps or duplicates"
+                );
+            }
+        }
+    }
+
+    // calculate listing bounds
     let start_version = start_version.into().unwrap_or(0);
     let end_version = end_version.into();
-    let version_prefix = format!("{start_version:020}");
-    let start_from = log_root.join(&version_prefix)?;
+    // start_from is log path to start listing from: the log root with zero-padded start version
+    let start_from = log_root.join(&format!("{start_version:020}"))?;
+    let list_end_version = match (log_tail.first(), end_version) {
+        // stop before log_tail or at requested end, whichever comes first
+        (Some(tail), Some(end)) => Some(end.min(tail.version.saturating_sub(1))),
+        // no end: stop just before log_tail starts
+        (Some(tail), None) => Some(tail.version.saturating_sub(1)),
+        // no tail: just respect requested end
+        (None, end) => end,
+    };
```

Comment:
note i used to do something fancier here with combinators but ultimately decided on `match` just for ease of reading.

Reply:
made some slight changes and if i rename closure to `head` instead of `first` then `fmt` keeps it on one line :)